manual: More on stores, building and mounting in the file system (#14699)
NixOS/nix: doc/manual/source/store/store-path.md

# Store Path and Store Directory

Nix's [store object] and [file system object] data models are minimal and abstract.
But to actually be used by software, store objects need to be made available through the operating system's file system.

This is done by exposing all the store objects in a single *[store directory][store directory path]*.

> Every entry in that directory has a *[store path base name]* as the name, and the file system object of one of the store objects in the store as the child file.

Reviewer: ```suggestion
Every entry in that directory is a *[store path base name]* pointing to a store object.
```

> **Note**
>
> The store directory defaults to `/nix/store`, but is in principle arbitrary.

## Store Path

A store path is the pair of a store directory path and a [store path base name].
It is rendered to a file system path as the concatenation of

- [Store directory] (typically `/nix/store`)
- Path separator (`/`)
- The [store path base name]

> **Example**
>
> ```
>   /nix/store/q06x3jll2yfzckz2bzqak089p43ixkkq-firefox-33.1
>   |--------| |------------------------------| |----------|
> store directory            digest                 name
> ```

When we have fixed a given store, or given store directory path (that all the stores in use share), the abstract syntax for store paths and the abstract syntax for store path base names coincide: the store directory path is known from context, so only the other two fields vary from one store path to the next.

## Exposing Store Objects in OS File Systems {#exposing}

Not every store can be accessed through the file system.
But if the store has a file system representation, the following should be true:

- The store directory path is canonical: no prefix of the path (i.e. path of the first *n* path segments) points to a symlink.

>   In other words, the store directory can be looked up from the store directory path without following any symlinks.

Reviewer 2: This isn't entirely correct, we have the `--allow-symlinked-store` which is no longer broken (though it has been for however many previous releases).

[store directory path]: #store-directory-path

## Store Path Base Name {#base-name}

Nix implements references to store objects as *store path base names*.

Think of a store path base name as an [opaque], [unique identifier]:
The only way to obtain a store path base name is by adding or building store objects.
A store path base name will always reference exactly one store object.

[opaque]: https://en.m.wikipedia.org/wiki/Opaque_data_type
[unique identifier]: https://en.m.wikipedia.org/wiki/Unique_identifier

Store path base names are pairs of

- A 20-byte [digest]{#digest} for identification
- A symbolic [name]{#name} for people to read

> **Example**
>
> - Digest: `q06x3jll2yfzckz2bzqak089p43ixkkq`
> - Name:   `firefox-33.1`

A store path base name is rendered to a string as the concatenation of

- Digest rendered in [Nix32](@docroot@/protocols/nix32.md), a variant of [base-32](https://en.wikipedia.org/wiki/Base32) (20 hash bytes become 32 ASCII characters)
- Hyphen (`-`)
- Name

> **Example**
>
> ```
> q06x3jll2yfzckz2bzqak089p43ixkkq-firefox-33.1
> |------------------------------| |----------|
>               digest                 name
> ```

> [Base32]: https://en.wikipedia.org/wiki/Base32

Reviewer 2: Prior mentions have `base-32` instead? Does this link work?

- Digest rendered in [Nix32], a variant of [Base32] (20 hash bytes become 32 ASCII characters)
- Hyphen (`-`)
- Name

> **Example**
>
> ```
> q06x3jll2yfzckz2bzqak089p43ixkkq-firefox-33.1
> |------------------------------| |----------|
>               digest                 name
> ```

[Nix32]: @docroot@/protocols/nix32.md
[Base32]: https://en.wikipedia.org/wiki/Base32

Exactly how the digest is calculated depends on the type of store object being referenced.
Store path digests are *supposed* to be opaque, and so for most operations, it is not necessary to know the details.
That said, the manual has a full [specification of store path digests](@docroot@/protocols/store-path.md).

## Store Directory Path

Every [Nix store] has a store directory path.
This is an absolute, lexically canonical (not containing any `..`, `.`, or similar) path which points to the directory where all store objects are to be found.

[Nix store]: ./index.md

> **Note**
>
> The store directory defaults to `/nix/store`, but is in principle arbitrary.

## Store Path

A store path is the pair of a store directory path and a [store path base name].
It is rendered to a file system path as the concatenation of

> - [Store directory] (typically `/nix/store`)

Reviewer 2: This link is broken

- [Store directory] (typically `/nix/store`)
- Path separator (`/`)
- The [store path base name]

> **Example**
>
> ```
>   /nix/store/b6gvzjyb2pg0kjfwrjmg1vfhh54ad73z-firefox-33.1
>   |--------| |------------------------------| |----------|
> store directory            digest                 name
> ```

When we have fixed a given store, or given store directory path (that all the stores in use share), the abstract syntax for store paths and the abstract syntax for store path base names coincide: the store directory path is known from context, so only the other two fields vary from one store path to the next.

## Exposing Store Objects in OS File Systems {#exposing}

Not every store can be accessed through the file system.
But if the store has a file system representation, the following should be true:

- The store directory path is canonical: no prefix of the path (i.e. path of the first *n* path segments) points to a symlink.
  (This is a separate condition in addition to "lexical canonicity", which is a property of just the path itself, whereas regular "canonicity" is an additional property about the path and the filesystem it navigates jointly.)

> - The store directory path in fact points to a directory.

Author: Hmm interesting point. I suppose I would say that seeing the directory externally sort of seeing an exposed implementation detail. The "proper" way to observe the files is only from inside the chroot.

> - The store directory path in fact points to a directory.

Reviewer 2: > The "proper" way to observe the files is only from inside the chroot.

Not really, `nix shell` does set up the mounts via `chrootHelper`.

> - The store directory path in fact points to a directory.

Author: Hmmm? I mean are you saying that `nix shell` doesn't put one in the chroot?

> - The store directory path in fact points to a directory.

Reviewer 2: No pivot_root happens, but an overlaysfs/bind mounts are set up depending on whether the store already exists at its logical paths. So I guess your original point stands.

> **Note**
>
> The store directory defaults to `/nix/store`, but is in principle arbitrary.

## Store Path

A store path is the pair of a store directory path and a [store path base name].
It is rendered to a file system path as the concatenation of

- [Store directory] (typically `/nix/store`)
- Path separator (`/`)
- The [store path base name]

> **Example**
>
> ```
>   /nix/store/q06x3jll2yfzckz2bzqak089p43ixkkq-firefox-33.1
>   |--------| |------------------------------| |----------|
> store directory            digest                 name
> ```

When we have fixed a given store, or given store directory path (that all the stores in use share), the abstract syntax for store paths and the abstract syntax for store path base names coincide: the store directory path is known from context, so only the other two fields vary from one store path to the next.

## Exposing Store Objects in OS File Systems {#exposing}

Not every store can be accessed through the file system.
But if the store has a file system representation, the following should be true:

- The store directory path is canonical: no prefix of the path (i.e. path of the first *n* path segments) points to a symlink.

>   (This is a separate condition in addition to "lexical canonicity", which is a property of just the path itself, whereas regular "canonicity" is an additional property about the path and the filesystem it navigates jointly.)

Reviewer 2: It's a pretty roundabout way of saying that symlinks are not resolved.

Author: I like the "canonical" vs "lexically canonical" distinction, but I added the other way of looking at it to help

- Suppose that the store directory (path) is `/foo/bar`.
  Then, `/foo/bar/b6gvzjyb2pg0kjfwrjmg1vfhh54ad73z-firefox-33.1` exists and is the file system object of that store object.

- Suppose that we don't know what the store directory path of the store is, but we do have a capability `storeDir` to the store directory on the file system.
  (This would be a "file descriptor" on Unix, or a "file handle" on Windows.)
  Then (using the Unix notation for this):
  ```
  openat(storeDir, "b6gvzjyb2pg0kjfwrjmg1vfhh54ad73z-firefox-33.1", O_NOFOLLOW)
  ```
  will succeed (so long as the file system object is not a symlink), and the yielded capability will point to the file system object of that store object.

  (The behavior for symlinks is harder to specify because of limitations in POSIX.)

## Relocating store objects

The inclusion of the store directory path in the full rendered store path means that it is not just derived from the referenced store object itself, but depends on the store that the store object is in.
(And actually, many of the ways of computing the digest also depend on the store directory path.

> So this is also true even just for store path base names, in general.)

Reviewer 2: Not sure what "many" refers to here. drvPath always depends on the store directory anyway.

Author: Should be better

  The permissions and other metadata for these files in the store directory is in the normal form described in [Exposing in OS file systems](./file-system-object/os-file-system.md).

The above properties mean that the following file accesses will work.
Suppose we have a store available on the file system per the above rules, and `b6gvzjyb2pg0kjfwrjmg1vfhh54ad73z-firefox-33.1` is the store path base name of a store object in that store.

- Suppose that the store directory (path) is `/foo/bar`.
  Then, `/foo/bar/b6gvzjyb2pg0kjfwrjmg1vfhh54ad73z-firefox-33.1` exists and is the file system object of that store object.

- Suppose that we don't know what the store directory path of the store is, but we do have a capability `storeDir` to the store directory on the file system.
  (This would be a "file descriptor" on Unix, or a "file handle" on Windows.)
  Then (using the Unix notation for this):
  ```
  openat(storeDir, "b6gvzjyb2pg0kjfwrjmg1vfhh54ad73z-firefox-33.1", O_NOFOLLOW)
  ```
  will succeed (so long as the file system object is not a symlink), and the yielded capability will point to the file system object of that store object.

>   (The behavior for symlinks is harder to specify because of limitations in POSIX.)

Reviewer 2: What limitations are we referring to?

Author: Oh I guess I could use `readlinkat`, rather than `openat` with `O_PATH` since that is Linux-only?
